File: backend/src/main/java/com/java/NBE4_5_1_7/global/util/LockManager.java
package com.java.NBE4_5_1_7.global.util;

import java.util.concurrent.TimeUnit;

import org.redisson.api.RLock;
import org.redisson.api.RedissonClient;

public class LockManager {

	private final RedissonClient redissonClient;

	public LockManager(RedissonClient redissonClient) {
		this.redissonClient = redissonClient;
	}

	public RLock getLock(String lockKey) {
		return redissonClient.getLock(lockKey);
	}

	public void acquireLock(RLock lock, long waitTime, long leaseTime) throws InterruptedException {
		if (!lock.tryLock(waitTime, leaseTime, TimeUnit.SECONDS)) {
			throw new RuntimeException("잠시 후 다시 시도해주세요.");
		}
	}

	public void releaseLock(RLock lock) {
		if (lock.isHeldByCurrentThread()) {
			lock.unlock();
		}
	}
}

File: backend/src/main/java/com/java/NBE4_5_3_7/global/util/LockManager.kt
package com.java.NBE4_5_3_7.global.util

import org.redisson.api.RLock
import org.redisson.api.RedissonClient
import java.util.concurrent.TimeUnit

class LockManager(
	private val redissonClient: RedissonClient
) {

	fun getLock(lockKey: String): RLock {
		return redissonClient.getLock(lockKey)
	}

	@Throws(InterruptedException::class)
	fun acquireLock(lock: RLock, waitTime: Long, leaseTime: Long) {
		if (!lock.tryLock(waitTime, leaseTime, TimeUnit.SECONDS)) {
			throw RuntimeException("잠시 후 다시 시도해주세요.")
		}
	}

	fun releaseLock(lock: RLock) {
		if (lock.isHeldByCurrentThread) {
			lock.unlock()
		}
	}
}
